Notes support

HookedBehemoth opened the pull request:
Closes https://github.com/zedeus/nitter/issues/646
My first draft so far with a few issues here and there to fix:

- [X] inlineStyles are applied to entire paragraphs (is it guaranteed that these don't overlap with links and media?)
- [X] hashtags/@ aren't linked to
- [X] unicode characters are badly handled in paragraphs with links¹
- [X] embeds have the wrong size² (see https://github.com/zedeus/nitter/issues/586#issuecomment-1159815907) (out of scope; directly embedding requires multiple requests)
- [X] ~~alt text on images isn't set~~

Not implemented (yet):
- [X] GIFs*
- [X] Lists*
- [X] Headers*
- [X] Twemoji**
*https://twitter.com/i/notes/1539382544856035328
**https://twitter.com/i/notes/1539522210976698373
<details>
<summary>Some screenshots</summary>
<img src="https://user-images.githubusercontent.com/22580720/176179793-c79419c1-b844-42d8-8848-1ab2755c66ae.png">

<p>(1)</p>
<img src="https://user-images.githubusercontent.com/22580720/176179588-e6bf5253-6d76-44c2-b0fd-024240b96ec8.png">

<p>(2)</p>
<img src="https://user-images.githubusercontent.com/22580720/176179973-f65926f0-b832-41fa-a314-ed2cb45301d5.png">
</details>

HookedBehemoth committed:
initial notes support

HookedBehemoth committed:
notes: use runeSubStr for ranges

HookedBehemoth committed:
notes: implement lists, headers and twemoji

HookedBehemoth committed:
notes: replace hashtags and mentions

HookedBehemoth committed:
notes: don't replace text in links

HookedBehemoth committed:
prevent karax from inserting whitespaces to fix wrapping

HookedBehemoth committed:
notes: proxy small images

HookedBehemoth committed:
add twemoji domain to whitelist

HookedBehemoth commented on src/types.nim:
```diff
@@ -18,6 +18,7 @@ type
     listMembers
     userRestId
     status
+    article
```
I'm not. How would I find out?

zedeus replied:
It's a little tricky, you basically have to call each endpoint with the same token and check the returned `x-rate-limit-remaining` header. If the article endpoint's rate limit goes down when you call some other endpoint, that endpoint's enum value here should be used.

HookedBehemoth committed:
notes: address review

HookedBehemoth committed:
notes: properly flush list

HookedBehemoth committed:
notes: use userRestId api token pool

HookedBehemoth committed:
notes: match twemoji size with font size

HookedBehemoth committed:
notes: ignore empty media

HookedBehemoth committed:
notes: render tweets inline

this resolves the issue of badly sized iframes and lowers the amount of
requests per client

HookedBehemoth committed:
notes: fix width to 600px

HookedBehemoth committed:
notes: support "gifs"

HookedBehemoth committed:
notes: treat tweets as mainTweet

zedeus commented on src/routes/notes.nim:
```diff
@@ -12,6 +12,18 @@ proc createNotesRouter*(cfg: Config) =
+        .await
+        .filterIt(it != nil)
+        .mapIt((it.tweet.id, it.tweet))
+        .toTable
```
This Rust-style solution is very non-idiomatic and slow in Nim since `filterIt` and `mapIt` fully create the lists instead of it being an iterator chain.

HookedBehemoth committed:
notes: fetch tweets in a nim idiomatic way

HookedBehemoth committed:
notes: return 404 on errors

HookedBehemoth committed:
notes: use absolute paths in style sheet to not apply to embeds

HookedBehemoth committed:
notes: don't emit paragraphs for atomic blocks

HookedBehemoth commented:
Should I bump the style sheet version parameter?

zedeus commented:
Yes, it invalidates the browser cache

HookedBehemoth committed:
notes: loop "gifs"

HookedBehemoth committed:
notes: bump stylesheet version

HookedBehemoth committed:
notes: support inlineStyles in the worst way possible

HookedBehemoth committed:
notes: fix inlineStyle length

HookedBehemoth committed:
notes: properly fix inlineStyle length this time

zedeus commented:
Replaced by https://github.com/zedeus/nitter/commit/35882ed88d422b1355b66a1ff8c1144bffdc7bdf, thanks!
